### PR TITLE
Upgrade to go 1.21.9

### PR DESCRIPTION
**Pull request by friedrichg**
**What this PR does**: Upgrades go to 1.21.9. Includes fixes for CVE-2023-45288, CVE-2024-24783, CVE-2023-45290, CVE-2023-45289

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/cortexproject/cortex
 
-go 1.21
+go 1.21.9
```
[the remark this answers is not in the document]
Comment (friedrichg) [reply go.mod:3]: This is the context https://github.com/golang/go/issues/62278 
codeql is failing https://github.com/cortexproject/cortex/security/code-scanning/tools/CodeQL/status/configurations/actions-FZTWS5DIOVRC653POJVWM3DPO5ZS65DFON2C2YTVNFWGILLEMVYGY33ZFZ4W23A/42065a090d6e5bc992b82fafa11d0bd003344f9e8baec9c2f02bffa068a55410

[the remark this answers is not in the document]
Comment (yeya24) [reply go.mod:3]: Is there a way to disable that check in codeql? I think both Thanos and Prometheus stays `go 1.21`.
https://github.com/thanos-io/thanos/blob/main/go.mod#L3 and https://github.com/prometheus/prometheus/blob/main/go.mod#L3

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,7 +27,7 @@
 * [ENHANCEMENT] Logging: Added new options for logging HTTP request headers: `-server.log-request-headers` enables logging HTTP request headers, `-server.log-request-headers-exclude-list` allows users to specify headers which should not be logged. #5744
 * [ENHANCEMENT] Query Frontend/Scheduler: Time check in query priority now considers overall data select time window (including range selectors, modifiers and lookback delta). #5758
 * [ENHANCEMENT] Querier: Added `querier.store-gateway-query-stats-enabled` to enable or disable store gateway query stats log. #5749
-* [ENHANCEMENT] Upgrade to go 1.21.6. #5765
+* [ENHANCEMENT] Upgrade to go 1.21.9 #5765
```
[the remark this answers is not in the document]
Comment (yeya24) [reply CHANGELOG.md:30]: Or mention #5879 in the same entry

**Comment by yeya24** [comment]
Any reason for the go version upgrade? Is this related to security issues?

**Pull request merged**